HACK: misc hacks for transparent sharded list/watch

stevekuznetsov opened the pull request:
Transparent sharding needs to occur after a client has passed auth{n,z}
for the primary replica they contact. This means that kcp needs to
inject a handler in between those chains and the gorestful delegates.
Right now the simplest way to do that is to just open up the option.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
HACK: allow configuring the handler chain from options

Transparent sharding needs to occur after a client has passed auth{n,z}
for the primary replica they contact. This means that kcp needs to
inject a handler in between those chains and the gorestful delegates.
Right now the simplest way to do that is to just open up the option.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

davidfestal reviewed (commented):
Having an insight about how you use / call this new option would help reviewing.

stevekuznetsov commented:
We only work with the options in kcp and I need to change the handler chain - the entire transformation from the raw options struct to the apiserver happens within these calls, so I exposed the knob I needed at the options level.

stevekuznetsov committed:
HACK: opt kcp controllers out of multi-cluster

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
HACK: export continue token impl

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
HACK: expose unstructured creator and helpers

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
HACK: rest: request: allow debugging incoming and outgoing requests

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
HACK: tablewriter: allow a default type-less handler

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
HACK: make e2e pass against kcp

There are a couple issues with the current e2e, including some weird
client config munging that breaks our setup as well as a lot of
assumptions that the cluster under test has Pods, Nodes, etc. We need to
turn all of this off so that the e2e is meaningful.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
HACK: prefix cluster names with external host

We need to have globally unique logical cluster names in order to be
able to meaningfully delegate requests to the cross-shard proxy.
Furthermore, until a source of truth exists for the mapping between
Workspaces and shards, we can get a fascimile of this by encoding the
serving address of the kcp shard into the workspace name and making sure
we pass around kubeconfigs with correct identifiers.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
server: move kcp-specific cluster handler to kcp

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
horrible hack: why?

make lists when we need lists... but why ?

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
e2e: add a kcp suite

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
HACK: add generated bits for go 1.17

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
client-go, client-gen: support cluster explicitly

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
clients: regenerate

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

ncdc commented:
LGTM to allow Steve to make progress in kcp
